keps/sig-api-machinery/4460-per-request-deadline/README.md

```diff
@@ -0,0 +1,1055 @@
+		return serialFinisher(ctx, fn)
+	}
+	return finishRequest(ctx, fn, postTimeoutLoggerWait, logPostTimeoutResult)
+}
+```
+
+
+### Monitor Post-Timeout Activity:
+
+Upon timeout of a request, the `timeout` filter today provides with the following:
+
+- increment the merics counter `request_post_timeout_total` by 1 for every 
+  request that has timed out.
+- print a line to the log
+
+```
+	err := fmt.Errorf("post-timeout activity - time-elapsed: %s, %v %q result: %v",
+		time.Since(timedOutAt), req.Method, req.URL.Path, result)
+	utilruntime.HandleError(err)
+```
+
+> After a request times out, the asynchronous finisher will spin up a new 
+> goroutine and wait up to `5m` for the callee/child goroutine to return.
+>
+> The timeout filter does the same thing, but waits indefinitely until the
+> callee goroutine returns.
+
+
+With the `timeout` filter removed, we need to build a new mechanism to maintain parity.
+
+##### Global Synchronized Hashmap
+
+We can setup a `PostStartHook` that will detect how long a requst handler
+goroutine is active after the timeout happens, and provide the same observability
+as today.
+
+One simple approach for alpha, is to maintain a global synchronized hashmap of 
+the requests that are being tracked by the `WithPerRequestTimeout` filter.
+```
+type Tracker interface {
+	Enter(key *http.Request, data *RequestData)
+	Leave (key *http.Request)
+}
+
+type RequestData struct {
+	StartedAt      time.Time
+	DeadlineAt     time.Time
+	RequestURI     string
+	RequestAuditID string
+}
+
+type tracker struct{
+	inflight map[*http.Request]*RequestData
+}
+
+// Inside the WithPerRequestTimeout filetr
+
+func(w http.ResponseWriter, req *http.Request) {
+...
+	data := &RequestData{StartedAt: startedAt, DeadlineAt: deadlineAt, RequestURI: "", AuditID: ""}
+	tracker.Enter(req, data)
+	defer tracker.Leave(req)
+
+	handler.ServeHTTP(w, req)
+...
+```
+
+The `PostStartHook` goroutine will wake up every `5m`, and perform the following operation:
+
+- Iterate through the map entries 
+- If an entry is resident for more than `5m` after the specified `DeadlineAt` then
+  it is reported in the metrics/logs
+- If an entry is resident for more than `10m`, we remove it from the hashmap so the
+  list does not grow infinitely.
+  
+
+The memory cost of the above hashmap is to the `O(number of non long-running requests in flight)`.
+  
+One tradeoff of this global hashmap is the locking contention present due to
+in-flight requests trying to access the synchronized map concurrently. But it
+avoids creating a new goroutine for each request.
+
+##### Post-Tomeout Goroutine
+
+Inside the `WithPerRequestTimeout` filter we can kick off a goroutine to monitor
+post-timeout activity, but this would include all non long-running request(s).
+
+Ideally, we want to spin up goroutine(s) for only those request(s) that time out.
+How do we get notified when the `Context` of a request times out, especially when
```

Comment:
Is it possible to watch the writes (similar to audit today) and simply start a goroutine on every 504 that lasts 5m?

Reply:
I think it's similar to this, if I am not misunderstanding
```
// One option is to tap into the `Write`, and `FlushError` methods of the `ResponseWriter` object:

type timeoutWriter struct {
	http.ResponseWriter
	once sync.Once
}

func (tw *timeoutWriter) Write(p []byte) (int, error) {
	n, err := tw.ResponseWriter.Write(p)
	if errors.IsTimeout(err) {
		tw.once.Do(func() {
			// kick off a goroutine to monitor
		})
	}
	return n, err
}

func (tw *timeoutWriter) FlushError() error {
	err := tw.ResponseWriter.FlushError()
	if errors.IsTimeout(err) {
		tw.once.Do(func() {
			// kick off a goroutine to monitor
		})
	}	
}
```

This only works for request(s) that are trying to write to the `ResponseWriter`
object after the timeout happens, I have proposed a solution using `context.AfterFunc`